traits/trait_dict_object.py

```diff
@@ -111,224 +400,142 @@ def __init__(self, trait, object, name, value):
         if trait.has_items:
             self.name_items = name + "_items"
 
-        if len(value) > 0:
-            dict.update(self, self._validate_dic(value))
+        super().__init__(value, key_validator=self._key_validator,
+                         value_validator=self._value_validator,
+                         notifiers=[self.notifier])
 
-    def _send_trait_items_event(self, name, event, items_event=None):
-        """ Send a TraitDictEvent to the owning object if there is one.
-        """
-        object = self.object()
-        if object is not None:
-            if items_event is None and hasattr(self, "trait"):
-                items_event = self.trait.items_event()
-            object.trait_items_event(name, event, items_event)
+    def _key_validator(self, key):
+        """ Calls the trait's key_trait.handler.validate.
 
-    def __deepcopy__(self, memo):
-        id_self = id(self)
-        if id_self in memo:
-            return memo[id_self]
+        Parameters
+        ----------
+        key : A hashable object.
+            The key to validate.
 
-        memo[id_self] = result = TraitDictObject(
-            self.trait,
-            lambda: None,
-            self.name,
-            dict(copy.deepcopy(x, memo) for x in self.items()),
-        )
+        Returns
+        -------
+        validated_key : A hashable object.
+            The validated key.
 
-        return result
+        Raises
+        ------
+        TraitError
+            If the validation fails.
+        """
 
-    def __setitem__(self, key, value):
-        trait = getattr(self, "trait", None)
-        if trait is None:
-            dict.__setitem__(self, key, value)
-            return
+        trait = getattr(self, 'trait', None)
+        object_ref = getattr(self, 'object', None)
+        if object_ref is None or trait is None:
```

Comment:
Could we add a comment explaining why this check is necessary? And similarly for the `_key_validator`? (I'm currently failing to see why it _is_ necessary, so I don't know what that comment would say ...)

Reply:
`__setstate__` sets the `trait` attribute to `None` when de-serialized, so `trait.key_trait.handler.validate` in the following lines, would cause an error. But I guess the `object_ref is None` check can be removed, but `object is None` check should be added.